### ISSUE TITLE
[Earthworm Jim 2] Sprite glitches

### ISSUE THREAD

**Issue by Aethyrion** [FCare/Kronos#473]
Kronos_1.6.0 (standalone & core libretro)
Windows 10

Version US d'Earthworm Jim 2, bug graphique sur les sprites apparaissant par intermittence.

![EJ2](https://user-images.githubusercontent.com/10311145/56890764-afd46f00-6a7a-11e9-99c8-bc02c8f96173.jpg)

Le problème n'apparaît pas sur la version PAL Europe.

**Comment by fafling** [comment]
Le sprite s'affiche correctement dans l'écran debug du VDP1.
![Capture_earthworm_jim3](https://user-images.githubusercontent.com/7624262/56933886-f9a86e00-6ae9-11e9-9e96-7949246d9e8c.PNG)
![Capture_earthworm_jim4](https://user-images.githubusercontent.com/7624262/56933891-fdd48b80-6ae9-11e9-8461-629e2be9126b.PNG)

Ressemble à ce problème : #412 
Le seul point commun avec NBA et NFL c'est le format 8 bits des textures.
On aurait pu penser à un problème d'implémentation de l'end code, utilisé par Earthworm Jim et NBA, mais NFL ne s'en sert pas.

**Comment by fafling** [comment]
Pas de différence dans la commande de dessin de Earthworm Jim en version PAL qui n'a effectivement pas ce problème.

**Comment by FCare** [comment]
Il faut savoir que l'image qu'on voit dans les fenetres DebugVDP1 sont
uniquement calculées par le moteur de rendu Soft, contrairement a l'image
du jeu qui est rendue en OpenGL (aligné les deux, c'est dans la todo list).

Par contre, ce qu'on peut en deduire:
- Ca n'est pas lié à l'implementation OpenGL (vu que le probleme est
parfois affiché par le moteur Soft).
- Ca n'est pas lié au multi-thread de vidogl.c
- Ca vient effectivement de la recuperation du character lui-meme.

Les premieres verifications dans le code montrent le meme resultat.

Graphiquement, on dirait un stride faux mais les verifications a
l'execution ne montre rien dans ce cas la. Ca pourrait etre egalement un
mauvais calcul de la largeur du sprite...

Pas evident celui la :-)

Le mar. 30 avr. 2019 à 01:57, nicolas-charrier <notifications@github.com> a
écrit :

**Comment by FCare** [comment]
En ajoutant un printf, j'arrive a faire disparaitre les glitches. C'est
donc un depassement memoire qque part..

**Comment by FCare** [comment]
Ok, j'ai trouvé la raison... Pour demarrer le rendu du VDP1, on doit normalement remplir une liste de commandes apres avoir copié les textures. Sauf que pour augmenter les perfs, certains jeux (dont earthrworm jim 2) demande l'execution de la table de commandes VDP1 avant qu les textures soient effectivement chargées. Dans l'emulateur, des qu'on recoit la commande, on dessine directement toutes les commandes. La solution consiste a executer les commandes a la vitesse simulée du VDP1.

**Comment by FCare** [comment]
En y regardant de pplus pret. La valeur PTMR est ecrite a 1 la plupart du temps en fin de trame, vers la ligne 253. Sauf une fois de temps en temps, dans la meme trame, on a un ordre de dessiner en ligne 7 et en ligne 253 alors qu'on avait rien a la trame precedente. Il semble donc que ca soit un pb de synchro.
Or ni l'EDSR, ni l'interruption HBlank ou Vblank ne sont utilisés pour la synchro. Il se peut donc que ce soit un du a un pb de synchro avec le son...

**Comment by fafling** [comment]
D'après la doc du VDP1, c'est l'accès à la VRAM par le SCU qui est prioritaire sur le VDP1, donc s'il y a un transfert en cours vers la VRAM, le VDP1 ne devrait pas être en mesure de commencer à dessiner avant que le transfert soit terminé, même si on met à jour son registre PTMR avant la fin de ce transfert.

**Comment by FCare** [comment]
j'ai essayé d'implémenter. Ca ameliore un peu mais ca se met a clignoter... Je reviendrais sur ce bug plus tard...

**Comment by fafling** [comment]
C'est peut-être ce genre de soucis qu'il y a dans Doom vu qu'il enchaîne les batchs de commandes dans la même frame. Si le registre PTMR est n'est pas mis à jour quand il faut, il y a des batchs qui peuvent être partiellement ou complètement écrasés avant que le VDP1 ne puisse les dessiner, non ?

**Comment by FCare** [comment]
Oui, le probleme est le meme. Ce n'est pas clair pour moi si ces jeux
terminent d'ecrire les textures dans la VRAM apres avoir demander
l'execution de la liste des commandes VDP1 ou si c'est un pb dans le
pipeline de rendu. Il semble bien qu'en jouant avec le delai de prise en
compte du PTMR, ca semble regler le pb. Sauf que ca casse d'autres jeux
comme Sega rally qui utilise des liste de commandes VDP1 tres longues.
Dans le moteur actuel, des que le PTMR est ecrit a 1, on execute la liste
des commandes, avec la VRAM dans l'etat au moment de l'ecriture du PTMR. Et
on l'execute ensuite aux trames suivantes, au meme numero de ligne....

Le lun. 6 mai 2019 à 23:52, nicolas-charrier <notifications@github.com> a
écrit :

**Comment by fafling** [comment]
Je pense que le fonctionnement typique, ça doit être de mettre a jour le PTMR lors de l'interrupt de fin de transfert du SCU DMA des commandes/textures d'un batch vers la VRAM.

**Comment by fafling** [comment]
En WIP du 24/08, le problème est présent dans les 2 noyaux, OpenGL et compute shader.

**Comment by BenjaminSiskoo** [comment]
Wip du 27/10/2019, problème toujours présent

**Comment by fafling** [comment]
Toujours présent en WIP du 06/12 dans les 2 noyaux.

**Comment by BenjaminSiskoo** [comment]
Pareil avec la Wip du 15/01/2020

**Comment by BenjaminSiskoo** [comment]
@fafling il semble que ce soit corrigé en CS/OpenGl, tu peux confirmer ?

**Comment by fafling** [comment]
Sur le sprite d'Earthworm Jim, je ne vois plus le problème, mais il est toujours présent sur d'autres sprites dans les 2 noyaux :
- Sur la boule qui roule :
![image](https://user-images.githubusercontent.com/7624262/96655617-39968d00-133e-11eb-9353-61ea1795038e.png)
- Sur le perso qui frappe Earthworm Jim :
![image](https://user-images.githubusercontent.com/7624262/96655851-c7727800-133e-11eb-9b0a-fe20a7d32803.png)

Save state juste avant ce point (aller vers la gauche, version US) :
[Earthworm_jim_sprite_corrompu.zip](https://github.com/FCare/Kronos/files/5412251/Earthworm_jim_sprite_corrompu.zip)

**Comment by FCare** [comment]
A retester.

**Comment by fafling** [comment]
Dans la dernière WIP, en noyau CS, la corruption n'est plus là sur le perso qui frappe Earthworm Jim, mais il peut apparaître en double (il n'y a rien qui corresponde à un dessin en double dans la liste de commande à ce moment-là) :
![image](https://user-images.githubusercontent.com/7624262/200137385-bc85a8c3-fa8f-4172-b92b-66e3847c63f4.png)

En OpenGL, on  voit non seulement cet affichage en double, mais la corruption est toujours là :
![image](https://user-images.githubusercontent.com/7624262/200137529-253ee7af-920a-4aaf-9f82-2a62cbcbb0c6.png)

Nouveau save state de la version US avant ce point (aller à gauche), car l'ancien ne marche plus :
[Earthworm_jim_sprite_corrompu_v2.3.1_20221105.zip](https://github.com/FCare/Kronos/files/9944579/Earthworm_jim_sprite_corrompu_v2.3.1_20221105.zip)
